docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,244 @@
+# EP: Definition of the Fleet and Device API spec
+<!-- this format is inspired by the K8S KEP format https://raw.githubusercontent.com/kubernetes/enhancements/master/keps/NNNN-kep-template/README.md -->
+## Summary
+
+The Device API, and Fleet API, and the dynamics of managing single fleetless
+devices as well as fleets of devices is a fundamental piece of flightctl.
+
+The Fleet API must serve as a way to manage groups of devices, providing enough
+flexibility to prevent unnecessary duplication of fleets, while still allowing
+for the management of devices in a way that is meaningful to the administrator.
+
+
+## Motivation
+
+We need agreement on the API design direction and the semantics of the API to be able to
+implement and grow the flightctl service.
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+* Define the format and behavior of the Device and Fleet APIs
+
+* Allow the management of individual fleetless devices on a smaller scale,
+  as an integration mechanism for services like AAP and ACM.
+
+* Allow the management of fleets of devices on a larger scale, via gitops or
+  API.
+
+* Define templating mechanism for the device spec template configuration in fleets,
+  based on labels and other metadata.
+
+### Non-Goals
+
+* Not defining the whole API in exhaustive form, this document is proposing changes
+  to the existing MVP implementation.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+### Device API
+
+The device API is divided into multiple endpoints, some user facing, and other device/facing,
+this design facilitates cleaner RBAC control of the endpoint access.
+
+#### User/Admin facing
+
+The fronted user/admin facing API endpoint is found in `/device/{name}`,
+this endpoint can be accessed by the administrator or external systems to
+manage individual devices. When a device is managed by a fleet flightctl is responsible
+of maintaining the device spec in sync with the fleet spec, in this case write access is
+possible for the purpose of the modification of labels and other metadata.
+
+The proposed changes include:
+
+* Adding a managementMode field to the device.spec, this field can be set to "fleet" or "manual".
```

Comment:
I believe the idea is to make this enum for potential future extensibility. If we go this route, we shouldn't call it "fleet", though, as this'd be a layering violation: The Device API must not know anything about higher layer APIs like the Fleet API.

However, I'm not convinced we should have this field at all: It's a duplicate way for users to express the intent to have a device managed or not, as we already support this through the fleet label-selector / setting the "owner" on a device. Also, it's semantic isn't clear: What if the field was set to "fleet" but no fleet selected the device?

If we absolutely want another layer of protecting the user from themselves, we could use a "pauseUpdates" or "suspendManagement" or something field or so that expresses the purpose better. And then maybe it should be a label/annotation instead of a spec?

Reply:
I don't love magic labels.  Why not leave it in the spec?

Reply:
There are some reasons why "magic labels" aren't always great, i.e.: There is no clear contract of field in the API, there is a contract, yes... in documentation, but you cannot find about it in the openapi doc for example, or while writing code using a client library. On the other hand, it's great for filtering in this use case.

There are some comments @fzdarsky that explain why it makes more sense to leave it out of spec (writablity/access)

An alternative could be, if we don't want "magic labels" to use a separate top level field:

i.e.
```
metadata:
  - 

management:
   suspended: true

spec:
  -
status:

```

Reply:
> 1. A device being managed manually is different than a device that happens not to match any selectors at the moment.

In how far? From the Device API layer's perspective, a human operator or a workflow automation or a fleet automation... these are just API users.

> 2. Artificially changing a device's labels to (possibly temporarily) exclude it from a fleet is an awkward UX.

Yeah, I agree. Though IMHO not more than having to double-express an intent. However, I feel the intent "to (possibly temporarily) exclude if from a fleet" is different than "to switch between manually and fleet managed". The former is what I meant above with a label or annotation that indicate to the automation layer to exclude the device (effectively, it becomes part of the higher layer's API).

Reply:
Ok, I don't like my own proposal. management.suspended could be interpreted in a different way.. that regardless of what we write here... it will not propagate to the device... I am withdrawing that proposal. Back to magic labels for now

Reply:
This is transparent to the agent; the agent does not need to modify any behavior. It simply receives an expected state, calculated on the server side.